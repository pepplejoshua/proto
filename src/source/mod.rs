pub mod errors;
pub mod source;
