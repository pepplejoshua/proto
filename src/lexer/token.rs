use std::rc::Rc;

use crate::source::source::{SourceFile, SourceRef};

#[allow(dead_code, unused)]
#[derive(Debug, Copy, Clone, PartialEq, Eq)]
pub enum TokenType {
    // keywords
    Pub,
    Fn,
    Var,
    If,
    Else,
    For,
    In,
    Break,
    Continue,
    True,
    False,
    Return,
    Struct,
    Print,
    Println,
    Some,
    None,
    Defer,
    Const,
    Comptime,
    Use,
    Module,
    As,
    Root,

    // operators
    BackSlash,
    Plus,
    Minus,
    Star,
    Slash,
    Modulo,
    PlusAssign,
    MinusAssign,
    StarAssign,
    SlashAssign,
    ModuloAssign,
    Ampersand,
    Equal,
    NotEqual,
    Less,
    LessEqual,
    Greater,
    GreaterEqual,
    Assign,
    And,
    Or,
    Not,

    // special characters
    At,

    // delimiters
    LParen,
    RParen,
    LCurly,
    RCurly,
    LBracket,
    RBracket,
    Semicolon,
    Colon,
    DoubleColon,
    ColonAssign,
    Comma,
    Dot,
    BackTick,

    // complex tokens
    Integer,
    Decimal,
    Identifier,
    Character,
    String,
    InterpolatedStringFragment,
    Comment,

    // type tags
    I8,
    I16,
    I32,
    I64,
    U8,
    U16,
    U32,
    U64,
    F32,
    F64,
    Bool,
    Void,
    Char,
    Str,
    Type,

    // misc
    Underscore,
    QuestionMark,
    Eof,
}

#[allow(unused)]
#[derive(Debug, Clone)]
pub struct SrcToken {
    pub ty: TokenType,
    pub loc: Rc<SourceRef>,
}

#[allow(unused)]
impl SrcToken {
    pub fn get_source_ref(self: &SrcToken) -> Rc<SourceRef> {
        return self.loc.clone();
    }

    pub fn as_str(self: &SrcToken, src: &SourceFile) -> String {
        match self.ty {
            TokenType::As => "as".to_string(),
            TokenType::Module => "module".to_string(),
            TokenType::Use => "use".to_string(),
            TokenType::Pub => "pub".to_string(),
            TokenType::Comptime => "comptime".to_string(),
            TokenType::Fn => "fn".to_string(),
            TokenType::Const => "const".to_string(),
            TokenType::Var => "var".to_string(),
            TokenType::If => "if".to_string(),
            TokenType::Else => "else".to_string(),
            TokenType::For => "for".to_string(),
            TokenType::In => "in".to_string(),
            TokenType::Break => "break".to_string(),
            TokenType::Continue => "continue".to_string(),
            TokenType::True => "true".to_string(),
            TokenType::False => "false".to_string(),
            TokenType::Return => "return".to_string(),
            TokenType::Struct => "struct".to_string(),
            TokenType::Print => "print".to_string(),
            TokenType::Println => "println".to_string(),
            TokenType::Some => "some".to_string(),
            TokenType::None => "none".to_string(),
            TokenType::Defer => "defer".to_string(),
            TokenType::Const => "const".to_string(),
            TokenType::BackSlash => "\\".to_string(),
            TokenType::Plus => "+".to_string(),
            TokenType::Minus => "-".to_string(),
            TokenType::Star => "*".to_string(),
            TokenType::Slash => "/".to_string(),
            TokenType::Modulo => "%".to_string(),
            TokenType::PlusAssign => "+=".to_string(),
            TokenType::MinusAssign => "-=".to_string(),
            TokenType::StarAssign => "*=".to_string(),
            TokenType::SlashAssign => "/=".to_string(),
            TokenType::ModuloAssign => "%=".to_string(),
            TokenType::Ampersand => "&".to_string(),
            TokenType::Equal => "==".to_string(),
            TokenType::NotEqual => "!=".to_string(),
            TokenType::Less => "<".to_string(),
            TokenType::LessEqual => "<=".to_string(),
            TokenType::Greater => ">".to_string(),
            TokenType::GreaterEqual => ">=".to_string(),
            TokenType::Assign => "=".to_string(),
            TokenType::And => "and".to_string(),
            TokenType::Or => "or".to_string(),
            TokenType::Not => "!".to_string(),
            TokenType::At => "@".to_string(),
            TokenType::LParen => "(".to_string(),
            TokenType::RParen => ")".to_string(),
            TokenType::LCurly => "{".to_string(),
            TokenType::RCurly => "}".to_string(),
            TokenType::LBracket => "[".to_string(),
            TokenType::RBracket => "]".to_string(),
            TokenType::Semicolon => ";".to_string(),
            TokenType::Colon => ":".to_string(),
            TokenType::DoubleColon => "::".to_string(),
            TokenType::ColonAssign => ":=".to_string(),
            TokenType::Comma => ",".to_string(),
            TokenType::Dot => ".".to_string(),
            TokenType::BackTick => "`".to_string(),
            TokenType::Integer
            | TokenType::Decimal
            | TokenType::Identifier
            | TokenType::InterpolatedStringFragment
            | TokenType::Comment
            | TokenType::String
            | TokenType::Character => {
                let snippet = &src.text[self.loc.flat_start..self.loc.flat_end];
                snippet.to_string()
            }
            TokenType::I8 => "i8".to_string(),
            TokenType::I16 => "i16".to_string(),
            TokenType::I32 => "i32".to_string(),
            TokenType::I64 => "i64".to_string(),
            TokenType::U8 => "u8".to_string(),
            TokenType::U16 => "u16".to_string(),
            TokenType::U32 => "u32".to_string(),
            TokenType::U64 => "u64".to_string(),
            TokenType::F32 => "f32".to_string(),
            TokenType::F64 => "f64".to_string(),
            TokenType::Bool => "bool".to_string(),
            TokenType::Void => "void".to_string(),
            TokenType::Char => "char".to_string(),
            TokenType::Str => "str".to_string(),
            TokenType::Type => "type".to_string(),
            TokenType::Underscore => "_".to_string(),
            TokenType::QuestionMark => "?".to_string(),
            TokenType::Eof => "\0".to_string(),
        }
    }
}
