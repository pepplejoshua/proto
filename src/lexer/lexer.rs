use super::token::Token;
use crate::source::{errors::LexError, source::SourceFile};

#[allow(dead_code)]
pub struct Lexer {
    pub src: SourceFile,
    last_lexed: Option<Token>,
    queue: Vec<Token>,
}

#[allow(dead_code)]
impl Lexer {
    pub fn new(src: SourceFile) -> Lexer {
        Lexer {
            src,
            last_lexed: None,
            queue: vec![],
        }
    }

    // skip all whitespace characters
    fn skip_whitespace(&mut self) {
        loop {
            let c = self.src.cur_char();
            if c.is_whitespace() {
                self.src.next_char();
                continue;
            } else {
                break;
            }
        }
    }

    fn lex_comment(&mut self) -> Result<Token, LexError> {
        let start = self.src.cur_char();
        let mut content = String::from(start);
        let mut span = self.src.get_ref();
        // skip both '/' characters
        content.push(self.src.next_char());
        span = span.combine(self.src.get_ref());
        self.src.next_char();

        while !self.src.is_eof() {
            let cur = self.src.cur_char();
            span = span.combine(self.src.get_ref());
            self.src.next_char();
            if cur != '\n' {
                content.push(cur);
                continue;
            } else {
                break;
            }
        }
        Ok(Token::SingleLineComment(span, content))
    }

    // try to lex the next possible token
    pub fn next_token(&mut self) -> Result<Token, LexError> {
        if !self.queue.is_empty() {
            let tok = self.queue.remove(0);
            self.last_lexed = Some(tok.clone());
            return Ok(tok);
        }

        while self.src.cur_char().is_whitespace() {
            // TODO: understand the ramifications of using newlines instead of ;
            // we care about newline characters since they take the place
            // of semicolons in the grammar
            // if self.src.cur_char() == '\n' {
            //     let newline_loc = self.src.get_ref();
            //     self.src.next_char();
            //     return Ok(Token::Newline(newline_loc));
            // }
            // skip all preceding whitespace
            self.skip_whitespace();
            // process comments
            // eventually, it'll get documentation comments
            // for now, it just skips single line comments
            // self.process_comment();
        }

        // get the current character
        let c = self.src.cur_char();

        // if we reached the end of the file, return EOF
        if c == '\0' {
            return Ok(Token::Eof(self.src.get_ref()));
        }
        let maybe_token = match c {
            '`' => self.lex_interpolated_string(),
            // operators
            '+' | '-' | '*' | '/' | '%' | '!' | '=' | '<' | '>' | '(' | ')' | '{' | '}' | '['
            | ']' | ',' | '.' | ':' | ';' | '@' | '|' | '&' | '?' => self.lex_operator(),
            // numbers
            _ if c.is_ascii_digit() => self.lex_number(),
            // identifiers | keywords
            _ if c.is_alphabetic() || c == '_' => self.lex_potential_identifier(),
            // ' is used to start characters
            '\'' => self.lex_char(),
            // " is used to start strings
            '"' => self.lex_string(),
            // invalid character
            _ => {
                let mut err_ref = self.src.get_ref();
                err_ref.end_col += 1;
                let err = Err(LexError::InvalidCharacter(err_ref));
                self.src.next_char(); // try to keep lexing
                err
            }
        };

        match &maybe_token {
            Ok(t) => self.last_lexed = Some(t.clone()),
            Err(_) => self.last_lexed = None,
        }
        maybe_token
    }

    fn lex_interpolated_string(&mut self) -> Result<Token, LexError> {
        let mut span = self.src.get_ref();
        self.src.next_char(); // consume initial backtick
        span = span.combine(self.src.get_ref());
        let bt = Token::BackTick(span.clone());
        let mut parts = vec![];

        let mut buf = String::new();
        let mut buf_span = self.src.get_ref();
        while !self.src.is_eof() {
            let c = self.src.cur_char();

            match c {
                '`' => {
                    span = self.src.get_ref();
                    self.src.next_char();
                    span = span.combine(self.src.get_ref());
                    if !buf.is_empty() {
                        parts.push(Token::InterpStrLiteral(buf_span.clone(), buf.clone()))
                    }
                    parts.push(Token::BackTick(span.clone()));
                    break;
                }
                '{' => {
                    if self.src.peek_char() == '{' {
                        // we can escape the { character and add it to the buf
                        self.src.next_char();
                        self.src.next_char();
                        buf.push('{');
                        buf_span = buf_span.combine(self.src.get_ref());
                        continue;
                    }

                    // store whatever we have in buf and reset it
                    span = span.combine(self.src.get_ref());
                    if !buf.is_empty() {
                        parts.push(Token::InterpStrLiteral(buf_span.clone(), buf.clone()));
                        buf.clear();
                    }

                    let mut lcurly_span = self.src.get_ref();
                    self.src.next_char();
                    lcurly_span = lcurly_span.combine(self.src.get_ref());
                    let lcurly = Token::LCurly(lcurly_span);
                    parts.push(lcurly);

                    // from this point till we see a }, we will call next_token() to read
                    // regular token to us, instead of a string
                    let mut saw_rcurly = false;
                    while !self.src.is_eof() {
                        if let Ok(tok) = self.next_token() {
                            span = span.combine(tok.get_source_ref());
                            saw_rcurly = matches!(tok, Token::RCurly(_));
                            parts.push(tok);
                            if saw_rcurly {
                                break;
                            }
                        } else {
                            break;
                        }
                    }

                    if !saw_rcurly {
                        return Err(LexError::UnterminatedStringLiteral(span));
                    }

                    buf_span = self.src.get_ref();
                }
                _ => {
                    buf.push(c);
                    self.src.next_char();
                    buf_span = buf_span.combine(self.src.get_ref());
                }
            }
        }

        if self.src.is_eof() {
            return Err(LexError::UnterminatedStringLiteral(span));
        }

        // let span =
        self.queue = parts;
        Ok(bt)
    }

    // a multi line string fragment is a string fragment that is
    // preceded by "---". It goes till the end of the line
    fn lex_multi_line_string_fragment(&mut self) -> Result<Token, LexError> {
        let mut content = String::new();
        let mut span = self.src.get_ref();
        // skip the remaining 2 '-' characters out of the 3
        // the first one was already consumed by lex_operator()
        self.src.next_char();
        self.src.next_char();
        self.src.next_char();

        span = span.combine(self.src.get_ref());

        while !self.src.is_eof() {
            let cur = self.src.cur_char();
            span = span.combine(self.src.get_ref());
            self.src.next_char();
            if cur != '\n' {
                content.push(cur);
                continue;
            } else {
                break;
            }
        }
        Ok(Token::MultiLineStringFragment(span, content))
    }

    // lex a potential identifier
    // it might turn out to be either a:
    // - keyword
    // - identifier
    fn lex_potential_identifier(&mut self) -> Result<Token, LexError> {
        let mut id = String::new();
        let c_ref = self.src.get_ref();
        let cur = self.src.cur_char();
        id.push(cur);

        // read all characters that are alphanumeric or '_'
        let mut end_ref;
        loop {
            let c = self.src.next_char();
            end_ref = self.src.get_ref();
            if c.is_alphanumeric() || c == '_' {
                id.push(c);
            } else {
                break;
            }
        }

        let combined_ref = c_ref.combine(end_ref);
        // check if the identifier is a keyword
        match id.as_str() {
            "fn" => Ok(Token::Fn(combined_ref)),
            "let" => Ok(Token::Let(combined_ref)),
            "mut" => Ok(Token::Mut(combined_ref)),
            "if" => Ok(Token::If(combined_ref)),
            "else" => Ok(Token::Else(combined_ref)),
            "loop" => Ok(Token::Loop(combined_ref)),
            "while" => Ok(Token::While(combined_ref)),
            "true" => Ok(Token::True(combined_ref)),
            "false" => Ok(Token::False(combined_ref)),
            "break" => Ok(Token::Break(combined_ref)),
            "continue" => Ok(Token::Continue(combined_ref)),
            "return" => Ok(Token::Return(combined_ref)),
            "pub" => Ok(Token::Pub(combined_ref)),
            "print" => Ok(Token::Print(combined_ref)),
            "println" => Ok(Token::Println(combined_ref)),
            "some" => Ok(Token::Some(combined_ref)),
            "none" => Ok(Token::None(combined_ref)),

            "i8" => Ok(Token::I8(combined_ref)),
            "i16" => Ok(Token::I16(combined_ref)),
            "i32" => Ok(Token::I32(combined_ref)),
            "i64" => Ok(Token::I64(combined_ref)),
            "int" => Ok(Token::Int(combined_ref)),
            "u8" => Ok(Token::U8(combined_ref)),
            "u16" => Ok(Token::U16(combined_ref)),
            "u32" => Ok(Token::U32(combined_ref)),
            "u64" => Ok(Token::U64(combined_ref)),
            "uint" => Ok(Token::UInt(combined_ref)),
            "bool" => Ok(Token::Bool(combined_ref)),
            "char" => Ok(Token::Char(combined_ref)),
            "str" => Ok(Token::Str(combined_ref)),
            "type" => Ok(Token::Type(combined_ref)),
            "void" => Ok(Token::Void(combined_ref)),
            "struct" => Ok(Token::Struct(combined_ref)),
            "mod" => Ok(Token::Mod(combined_ref)),
            "_" => Ok(Token::Underscore(combined_ref)),
            _ => Ok(Token::Identifier(id, combined_ref)),
        }
    }

    // lex an operator
    fn lex_operator(&mut self) -> Result<Token, LexError> {
        let cur_ref = self.src.get_ref();
        let cur = self.src.cur_char();

        match cur {
            // arithmetic operators
            '+' => {
                self.src.next_char();
                Ok(Token::Plus(cur_ref.combine(self.src.get_ref())))
            }
            '-' => {
                self.src.next_char();
                Ok(Token::Minus(cur_ref.combine(self.src.get_ref())))
            }
            '*' => {
                self.src.next_char();
                Ok(Token::Star(cur_ref.combine(self.src.get_ref())))
            }
            '/' => {
                // if the next character is a '/', lex a comment
                if self.src.peek_char() == '/' {
                    return self.lex_comment();
                }
                self.src.next_char();
                Ok(Token::Slash(cur_ref.combine(self.src.get_ref())))
            }
            '%' => {
                self.src.next_char();
                Ok(Token::Modulo(cur_ref.combine(self.src.get_ref())))
            }

            '@' => {
                self.src.next_char();
                Ok(Token::At(cur_ref.combine(self.src.get_ref())))
            }

            // logical operators
            '!' => {
                // if the next character is a '=', return a NotEqual operator
                let c = self.src.peek_char();
                if c == '=' {
                    self.src.next_char();
                    self.src.next_char();
                    return Ok(Token::NotEqual(cur_ref.combine(self.src.get_ref())));
                }
                // otherwise, return a Not operator
                self.src.next_char();
                Ok(Token::Not(cur_ref.combine(self.src.get_ref())))
            }
            '&' => {
                // if the next character is a '&', return a And operator
                let c = self.src.peek_char();
                if c == '&' {
                    self.src.next_char();
                    self.src.next_char();
                    return Ok(Token::And(cur_ref.combine(self.src.get_ref())));
                }
                // otherwise, return an error
                Err(LexError::InvalidCharacter(
                    cur_ref.combine(self.src.get_ref()),
                ))
            }
            '|' => {
                // if the next character is a '|', return a Or operator
                let c = self.src.peek_char();
                if c == '|' {
                    self.src.next_char();
                    self.src.next_char();
                    return Ok(Token::Or(cur_ref.combine(self.src.get_ref())));
                }
                // otherwise, return an error
                Err(LexError::InvalidCharacter(
                    cur_ref.combine(self.src.get_ref()),
                ))
            }
            '=' => {
                // if the next character is a '=', return a Equal operator
                let c = self.src.peek_char();
                if c == '=' {
                    self.src.next_char();
                    self.src.next_char();
                    return Ok(Token::Equal(cur_ref.combine(self.src.get_ref())));
                }
                // otherwise, return a Assign operator
                self.src.next_char();
                Ok(Token::Assign(cur_ref.combine(self.src.get_ref())))
            }
            '<' => {
                // if the next character is a '=', return a LessEqual operator
                let c = self.src.peek_char();
                if c == '=' {
                    self.src.next_char();
                    self.src.next_char();
                    return Ok(Token::LessEqual(cur_ref.combine(self.src.get_ref())));
                }
                // otherwise, return a Less operator
                self.src.next_char();
                Ok(Token::Less(cur_ref.combine(self.src.get_ref())))
            }
            '>' => {
                // if the next character is a '=', return a GreaterEqual operator
                let c = self.src.peek_char();
                if c == '=' {
                    self.src.next_char();
                    self.src.next_char();
                    return Ok(Token::GreaterEqual(cur_ref.combine(self.src.get_ref())));
                }
                // otherwise, return a Greater operator
                self.src.next_char();
                Ok(Token::Greater(cur_ref.combine(self.src.get_ref())))
            }
            ':' => {
                self.src.next_char();
                Ok(Token::Colon(cur_ref.combine(self.src.get_ref())))
            }
            '(' => {
                self.src.next_char();
                Ok(Token::LParen(cur_ref.combine(self.src.get_ref())))
            }
            ')' => {
                self.src.next_char();
                Ok(Token::RParen(cur_ref.combine(self.src.get_ref())))
            }
            '{' => {
                self.src.next_char();
                Ok(Token::LCurly(cur_ref.combine(self.src.get_ref())))
            }
            '}' => {
                self.src.next_char();
                Ok(Token::RCurly(cur_ref.combine(self.src.get_ref())))
            }
            '[' => {
                self.src.next_char();
                Ok(Token::LBracket(cur_ref.combine(self.src.get_ref())))
            }
            ']' => {
                self.src.next_char();
                Ok(Token::RBracket(cur_ref.combine(self.src.get_ref())))
            }
            ',' => {
                self.src.next_char();
                Ok(Token::Comma(cur_ref.combine(self.src.get_ref())))
            }
            '.' => {
                self.src.next_char();
                Ok(Token::Dot(cur_ref.combine(self.src.get_ref())))
            }
            ';' => {
                self.src.next_char();
                Ok(Token::Semicolon(cur_ref.combine(self.src.get_ref())))
            }
            '?' => {
                self.src.next_char();
                Ok(Token::QuestionMark(cur_ref.combine(self.src.get_ref())))
            }
            _ => unreachable!("invalid operator character: '{}' at {:?}", cur, cur_ref),
        }
    }

    fn lex_string(&mut self) -> Result<Token, LexError> {
        let mut span = self.src.get_ref();
        let mut content = String::new();

        // read all characters until terminating "
        // if a new line is encountered, return an error
        // if a \ is encountered, read the next character
        while !self.src.is_eof() {
            let c = self.src.next_char();
            span = span.combine(self.src.get_ref());
            if c == '"' || c == '\n' {
                break;
            } else if c == '\\' {
                // if the next character is a ", add it to the content
                let c = self.src.next_char();
                span = span.combine(self.src.get_ref());
                if c == '"' {
                    content.push('"');
                } else {
                    // otherwise, add the escape character
                    content.push('\\');
                    content.push(c);
                }
            } else {
                content.push(c);
            }
        }

        // if the string is not terminated, return an error
        if self.src.is_eof() || self.src.cur_char() == '\n' {
            Err(LexError::UnterminatedStringLiteral(span))
        } else {
            self.src.next_char();
            span = span.combine(self.src.get_ref());
            Ok(Token::SingleLineStringLiteral(span, content))
        }
    }

    // try to lex a character
    fn lex_char(&mut self) -> Result<Token, LexError> {
        let mut span = self.src.get_ref();
        let mut content = String::new();

        // read single character
        let c = self.src.next_char();
        span = span.combine(self.src.get_ref());

        // if the next character is a ', return an error
        if c == '\'' {
            return Err(LexError::EmptyCharacterLiteral(span));
        } else if c == '\\' {
            // if the next character is a ', add it to the content
            let c = self.src.next_char();
            span = span.combine(self.src.get_ref());
            if c == '\'' {
                content.push('\'');
            } else {
                // otherwise, add the escape character
                content.push('\\');
                content.push(c);
            }
        } else {
            content.push(c);
        }

        // if the next character is not a ', return an error
        let c = self.src.next_char();
        span = span.combine(self.src.get_ref());
        if c != '\'' {
            return Err(LexError::UnterminatedCharacterLiteral(span));
        }

        self.src.next_char();
        span = span.combine(self.src.get_ref());
        Ok(Token::CharLiteral(span, content.chars().next().unwrap()))
    }

    fn lex_number(&mut self) -> Result<Token, LexError> {
        let mut number = String::new();
        let cur_ref = self.src.get_ref();
        let cur = self.src.cur_char();
        number.push(cur);

        // read all characters that are digits
        let mut end_ref;
        loop {
            let c = self.src.next_char();
            end_ref = self.src.get_ref();
            if c.is_ascii_digit() {
                number.push(c);
            } else if c == '_' {
                // ignore underscores
                continue;
            } else {
                break;
            }
        }

        let combined_ref = cur_ref.combine(end_ref);

        // TODO: if the next character is a '.', try to lex a float

        Ok(Token::NumberLiteral(number, combined_ref))
    }
}

#[cfg(test)]
#[allow(dead_code)]
#[derive(serde::Deserialize, serde::Serialize, Debug)]
struct LexerTestResult {
    stringified_tokens: Vec<String>,
}

#[cfg(test)]
#[test]
fn test_lexer() {
    insta::glob!("lexer_inputs/*.pr", |path| {
        // build the SourceFile from the proto file
        let path = path.to_str().unwrap().to_string();
        let src: SourceFile = SourceFile::new(path);

        // build the lexer
        let mut lexer = Lexer::new(src);

        let mut tokens = Vec::new();
        let mut res = LexerTestResult {
            stringified_tokens: Vec::new(),
        };
        loop {
            let token = lexer.next_token();
            if token.is_err() {
                panic!("error lexing: {token:?}");
            }
            let token = token.unwrap();
            if let Token::Eof(_) = token {
                break;
            }
            res.stringified_tokens.push(format!(
                "{} at {}",
                token.as_str(),
                token.get_source_ref().as_str()
            ));
            tokens.push(token);
        }
        insta::assert_yaml_snapshot!(res)
    });
}
