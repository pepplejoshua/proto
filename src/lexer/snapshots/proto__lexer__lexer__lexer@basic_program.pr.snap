---
source: src/lexer/lexer.rs
expression: res
input_file: src/lexer/lexer_inputs/basic_program.pr
---
stringified_tokens:
  - "fn at [start_line:0 start_col:0 end_line:0 end_col:2 flat_start:0 flat_end:2]"
  - "main at [start_line:0 start_col:3 end_line:0 end_col:7 flat_start:3 flat_end:7]"
  - "( at [start_line:0 start_col:7 end_line:0 end_col:8 flat_start:7 flat_end:8]"
  - ") at [start_line:0 start_col:8 end_line:0 end_col:9 flat_start:8 flat_end:9]"
  - "void at [start_line:0 start_col:10 end_line:0 end_col:14 flat_start:10 flat_end:14]"
  - "{ at [start_line:0 start_col:15 end_line:0 end_col:16 flat_start:15 flat_end:16]"
  - "let at [start_line:1 start_col:4 end_line:1 end_col:7 flat_start:21 flat_end:24]"
  - "x at [start_line:1 start_col:8 end_line:1 end_col:9 flat_start:25 flat_end:26]"
  - "= at [start_line:1 start_col:10 end_line:1 end_col:11 flat_start:27 flat_end:28]"
  - "42 at [start_line:1 start_col:12 end_line:1 end_col:14 flat_start:29 flat_end:31]"
  - "mut at [start_line:2 start_col:4 end_line:2 end_col:7 flat_start:36 flat_end:39]"
  - "y at [start_line:2 start_col:8 end_line:2 end_col:9 flat_start:40 flat_end:41]"
  - "= at [start_line:2 start_col:10 end_line:2 end_col:11 flat_start:42 flat_end:43]"
  - "10 at [start_line:2 start_col:12 end_line:2 end_col:14 flat_start:44 flat_end:46]"
  - "// Basic arithmetic at [start_line:4 start_col:4 end_line:4 end_col:23 flat_start:52 flat_end:71]"
  - "let at [start_line:5 start_col:4 end_line:5 end_col:7 flat_start:76 flat_end:79]"
  - "z at [start_line:5 start_col:8 end_line:5 end_col:9 flat_start:80 flat_end:81]"
  - "= at [start_line:5 start_col:10 end_line:5 end_col:11 flat_start:82 flat_end:83]"
  - "x at [start_line:5 start_col:12 end_line:5 end_col:13 flat_start:84 flat_end:85]"
  - "+ at [start_line:5 start_col:14 end_line:5 end_col:15 flat_start:86 flat_end:87]"
  - "y at [start_line:5 start_col:16 end_line:5 end_col:17 flat_start:88 flat_end:89]"
  - "y at [start_line:6 start_col:4 end_line:6 end_col:5 flat_start:94 flat_end:95]"
  - "*= at [start_line:6 start_col:6 end_line:6 end_col:8 flat_start:96 flat_end:98]"
  - "2 at [start_line:6 start_col:9 end_line:6 end_col:10 flat_start:99 flat_end:100]"
  - "// Reference example at [start_line:8 start_col:4 end_line:8 end_col:24 flat_start:106 flat_end:126]"
  - "let at [start_line:9 start_col:4 end_line:9 end_col:7 flat_start:131 flat_end:134]"
  - "p at [start_line:9 start_col:8 end_line:9 end_col:9 flat_start:135 flat_end:136]"
  - "= at [start_line:9 start_col:10 end_line:9 end_col:11 flat_start:137 flat_end:138]"
  - "& at [start_line:9 start_col:12 end_line:9 end_col:13 flat_start:139 flat_end:140]"
  - "x at [start_line:9 start_col:13 end_line:9 end_col:14 flat_start:140 flat_end:141]"
  - "mut at [start_line:10 start_col:4 end_line:10 end_col:7 flat_start:146 flat_end:149]"
  - "q at [start_line:10 start_col:8 end_line:10 end_col:9 flat_start:150 flat_end:151]"
  - "= at [start_line:10 start_col:10 end_line:10 end_col:11 flat_start:152 flat_end:153]"
  - "& at [start_line:10 start_col:12 end_line:10 end_col:13 flat_start:154 flat_end:155]"
  - "mut at [start_line:10 start_col:13 end_line:10 end_col:16 flat_start:155 flat_end:158]"
  - "y at [start_line:10 start_col:17 end_line:10 end_col:18 flat_start:159 flat_end:160]"
  - "// Match expression at [start_line:12 start_col:4 end_line:12 end_col:23 flat_start:166 flat_end:185]"
  - "match at [start_line:13 start_col:4 end_line:13 end_col:9 flat_start:190 flat_end:195]"
  - "y at [start_line:13 start_col:10 end_line:13 end_col:11 flat_start:196 flat_end:197]"
  - "{ at [start_line:13 start_col:12 end_line:13 end_col:13 flat_start:198 flat_end:199]"
  - "0 at [start_line:14 start_col:8 end_line:14 end_col:9 flat_start:208 flat_end:209]"
  - "=> at [start_line:14 start_col:10 end_line:14 end_col:12 flat_start:210 flat_end:212]"
  - "x at [start_line:14 start_col:13 end_line:14 end_col:14 flat_start:213 flat_end:214]"
  - ", at [start_line:14 start_col:14 end_line:14 end_col:15 flat_start:214 flat_end:215]"
  - "_ at [start_line:15 start_col:8 end_line:15 end_col:9 flat_start:224 flat_end:225]"
  - "=> at [start_line:15 start_col:10 end_line:15 end_col:12 flat_start:226 flat_end:228]"
  - "0 at [start_line:15 start_col:13 end_line:15 end_col:14 flat_start:229 flat_end:230]"
  - ", at [start_line:15 start_col:14 end_line:15 end_col:15 flat_start:230 flat_end:231]"
  - "} at [start_line:16 start_col:4 end_line:16 end_col:5 flat_start:236 flat_end:237]"
  - "} at [start_line:17 start_col:0 end_line:17 end_col:1 flat_start:238 flat_end:239]"
